File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        int r = matrix.length;
        int c = matrix[0].length;
        int[][] result = new int[r][c];

        // Step 1: Copy original matrix into result
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                result[i][j] = matrix[i][j];
            }
        }

        // Step 2: Find zeros and mark corresponding rows and columns
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (matrix[i][j] == 0) {
                    solve(result, i, j); // Modify result matrix
                }
            }
        }

        // Step 3: Copy result back to matrix (modifies in-place)
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                matrix[i][j] = result[i][j];
            }
        }
    }

    public void solve(int[][] result, int row, int col) {
        int r = result.length;
        int c = result[0].length;

        // Set entire column to 0
        for (int i = 0; i < r; i++) {
            result[i][col] = 0;
        }
        // Set entire row to 0
        for (int j = 0; j < c; j++) {
            result[row][j] = 0;
        }
    }
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.cpp
#include <vector>
using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int r = matrix.size();
        int c = matrix[0].size();
        vector<vector<int>> result = matrix;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (matrix[i][j] == 0) {
                    solve(result, i, j);
                }
            }
        }
        matrix = result;
    }
    
private:
    void solve(vector<vector<int>>& result, int row, int col) {
        int r = result.size();
        int c = result[0].size();
        for (int j = 0; j < c; j++) result[row][j] = 0;
        for (int i = 0; i < r; i++) result[i][col] = 0;
    }
};
